components/service/contile/src/main/java/mozilla/components/service/contile/ContileTopSitesProvider.kt

```diff
@@ -4,32 +4,70 @@
 
 package mozilla.components.service.contile
 
+import android.content.Context
+import android.util.AtomicFile
+import androidx.annotation.VisibleForTesting
 import mozilla.components.concept.fetch.Client
 import mozilla.components.concept.fetch.Request
 import mozilla.components.concept.fetch.isSuccess
 import mozilla.components.feature.top.sites.TopSite
 import mozilla.components.feature.top.sites.TopSitesProvider
 import mozilla.components.support.base.log.logger.Logger
 import mozilla.components.support.ktx.android.org.json.asSequence
+import mozilla.components.support.ktx.util.readAndDeserialize
+import mozilla.components.support.ktx.util.writeString
 import org.json.JSONException
 import org.json.JSONObject
+import java.io.File
 import java.io.IOException
+import java.util.Date
 
 internal const val CONTILE_ENDPOINT_URL = "https://contile.services.mozilla.com/v1/tiles"
+internal const val CACHE_FILE_NAME = "mozilla_components_service_contile.json"
+internal const val MINUTE_IN_MS = 60 * 1000
 
 /**
  * Provide access to the Contile services API.
  *
+ * @property context A reference to the application context.
  * @property client [Client] used for interacting with the Contile HTTP API.
+ * @property endPointURL The url of the endpoint to fetch from. Defaults to [CONTILE_ENDPOINT_URL].
+ * @property maxCacheAgeInMinutes Maximum time (in minutes) the cache should remain valid
+ * before a refresh is attempted. Defaults to -1, meaning no cache is being used by default.
  */
 class ContileTopSitesProvider(
-    private val client: Client
+    private val context: Context,
+    private val client: Client,
+    private val endPointURL: String = CONTILE_ENDPOINT_URL,
+    private val maxCacheAgeInMinutes: Long = -1
 ) : TopSitesProvider {
 
     private val logger = Logger("ContileTopSitesProvider")
 
+    private val diskCacheLock = Any()
+
+    /**
+     * Fetches from the top sites [endPointURL] to provide a list of provided top sites.
+     * Returns a cached response if [allowCache] is true and the cache is not expired
+     * (@see [maxCacheAgeInMinutes]).
+     *
+     * @param allowCache Whether or not the result may be provided from a previously cached
+     * response. Note that a [maxCacheAgeInMinutes] must be provided in order for the cache to be
+     * active.
+     * @throws IOException if the request failed to fetch any top sites.
+     */
     @Throws(IOException::class)
-    override suspend fun getTopSites(): List<TopSite.Provided> {
+    override suspend fun getTopSites(allowCache: Boolean): List<TopSite.Provided> {
+        val cachedTopSites = if (allowCache && !isCacheExpired()) {
+            readFromDiskCache()
+        } else {
+            null
+        }
+
+        if (!cachedTopSites.isNullOrEmpty()) {
+            return cachedTopSites
+        }
+
         return try {
             fetchTopSites()
         } catch (e: IOException) {
```

Comment:
One thing you're not doing here relative to the add-ons provider is falling back to the cache in case of network errors, which we should probably do if the cache is still valid (not expired)?

Reply:
Related: We will need to track the time we last successfully fetched so we know when to refresh. So we will likely need to return that info here as well?

Reply:
(or expose a method for it.)

Reply:
Alright we talked this through and given the requirement to have different foreground/background intervals. We will use `allowCache=false` for the background updater, which will run every 3 hours. We will use the cache `allowCache=true` for foreground updates and configure the cache max age to be one hour. /cc @gabrielluong

Reply:
So this is ready to land.